llvm/lib/CAS/OnDiskCASLogger.cpp

```diff
@@ -0,0 +1,223 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/CAS/OnDiskCASLogger.h"
+
+#include "llvm/ADT/StringSwitch.h"
+#include "llvm/ADT/Twine.h"
+#include "llvm/Config/llvm-config.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/raw_ostream.h"
+#if LLVM_ON_UNIX
+#include <sys/time.h>
+#include <unistd.h>
+#endif
+
+using namespace llvm;
+using namespace llvm::cas;
+using namespace llvm::cas::ondisk;
+
+// The version number in this log should be bumped if the log format is changed
+// in an incompatible way. It is currently a human-readable text file, so in
+// practice this would be if the log changed to binary or other machine-
+// readable format.
+static constexpr StringLiteral Filename = "/v1.log";
+
+OnDiskCASLogger::OnDiskCASLogger(raw_fd_ostream &OS, bool LogAllocations)
+    : OS(OS), LogAllocations(LogAllocations) {}
+
+OnDiskCASLogger::~OnDiskCASLogger() {
+  OS.flush();
+  delete &OS;
+}
+
+static bool isDisabledEnv(StringRef V) {
+  return StringSwitch<bool>(V)
+      .Case("0", true)
+      .CaseLower("no", true)
+      .CaseLower("false", true)
+      .Default(false);
+}
+
+Expected<std::unique_ptr<OnDiskCASLogger>>
+OnDiskCASLogger::openIfEnabled(const Twine &Path) {
+  const char *V = getenv("LLVM_CAS_LOG");
+  if (V && !isDisabledEnv(V)) {
+    int LogLevel = -1;
+    StringRef(V).getAsInteger(10, LogLevel);
+    return OnDiskCASLogger::open(Path, /*LogAllocations=*/LogLevel > 1 ? true
+                                                                       : false);
+  }
+  return nullptr;
+}
+Expected<std::unique_ptr<OnDiskCASLogger>>
+OnDiskCASLogger::open(const Twine &Path, bool LogAllocations) {
+  std::error_code EC;
+  SmallString<128> FullPath;
+  (Path + Filename).toVector(FullPath);
+
+  auto OS =
+      std::make_unique<raw_fd_ostream>(FullPath, EC, sys::fs::CD_OpenAlways,
+                                       sys::fs::FA_Write, sys::fs::OF_Append);
+  if (EC)
+    return createFileError(FullPath, EC);
+
+  // Buffer is not thread-safe.
+  OS->SetUnbuffered();
+
+  return std::unique_ptr<OnDiskCASLogger>(
+      new OnDiskCASLogger{*OS.release(), LogAllocations});
+}
+
+namespace {
+/// Helper to log a single line that adds the timestamp, pid, and tid. The line
+/// is buffered and written in a single call to write() so that if the
+/// underlying OS syscall is handled atomically so is this log message.
+class TextLogLine : public raw_svector_ostream {
+public:
+  TextLogLine(raw_ostream &LogOS) : raw_svector_ostream(Buffer), LogOS(LogOS) {
+    startLogMsg(*this);
+  }
+
+  ~TextLogLine() {
+    finishLogMsg(*this);
+    LogOS.write(Buffer.data(), Buffer.size());
```

Comment:
Maybe good to assert on the size to make sure it stays under a limit.

Reply:
I'm not sure the best way to handle this. Most of the log messages should be short and asserting would be fine, but there are a few that contain paths and possibly error messages and I don't want to constrain those. The longer log messages are also the ones that are only emitted infrequently like when opening/closing.

Reply:
Sure. I can accept that. Maybe interleaving log is not the end of the world for debugging.